[quant] Dynamic Linear module to use reduce_range

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39187 [quant] Enable per-channel quantization for quantized nn.LSTM
* #39186 [quant] Make reduce_range default for RNNs
* **#39125 [quant] Dynamic Linear module to use reduce_range**
* #39041 [quant] Add reduce_range argument for qlinear_dynamic

Summary:
switch to setting reduce_range to true for version > 3.
Models serialized with older state_dict will have version <=3 so will be run with reduce_range=false

Verified with backward compatibility tests (works with no changes to these tests)

Test Plan:
python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21769689](https://our.internmc.facebook.com/intern/diff/D21769689)

supriyar committed:
[quant] Dynamic Linear module to use reduce_range
Summary:
switch to setting reduce_range to true for version > 3.
Models serialized with older state_dict will have version <=3 so will be run with reduce_range=false

Verified with backward compatibility tests (works with no changes to these tests)

Test Plan:
python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on torch/nn/quantized/dynamic/modules/linear.py:
```diff
@@ -28,18 +28,25 @@ class Linear(nnq.Linear):
         torch.Size([128, 30])
     """
 
+    _version = 4
```
just for my own learning, the previous version was 3 even though it wasn't specified in this file?

supriyar replied:
Yeah, dynamic Linear inherits from nn.quantized.Linear which currently has the version set at 3. So I decided to keep an independent version in this file. In the future if we bump the version of nnq.Linear to 4 or higher, this should still work since we check for `self.version < 4` in forward to call the older API.
Alternately I could bump the version in nnq.Linear to 4. But that would involve more changes to that file and might be a bit confusing. Let me know if you think that is better.

vkuzo replied:
makes sense.  Maybe we can add a comment that this version is different from the parent class, since it's easy to miss?

vkuzo reviewed (approved):
lg

supriyar committed:
Update on "[quant] Dynamic Linear module to use reduce_range"


Summary:
switch to setting reduce_range to true for version > 3.
Models serialized with older state_dict will have version <=3 so will be run with reduce_range=false

Verified with backward compatibility tests (works with no changes to these tests)

Test Plan:
python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Dynamic Linear module to use reduce_range"


Summary:
switch to setting reduce_range to true for version > 3.
Models serialized with older state_dict will have version <=3 so will be run with reduce_range=false

Verified with backward compatibility tests (works with no changes to these tests)

Test Plan:
python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21769689](https://our.internmc.facebook.com/intern/diff/D21769689)

[ghstack-poisoned]
